File: Sliding Window/leetcode_438.py

# Problem link ===>>  https://leetcode.com/problems/find-all-anagrams-in-a-string/description/
# submission link == =>> https://leetcode.com/problems/find-all-anagrams-in-a-string/submissions/1590274065/

from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        cnt_p = [0] * 26
        cnt_w = [0] * 26

        for c in p:
            # Used ord() to get character ASCII values
            cnt_p[ord(c) - ord('a')] += 1

        k = len(p)
        n = len(s)

        # Initialize the first window
        for i in range(k):
            if i < n:  # To handle case when k > n
                cnt_w[ord(s[i]) - ord('a')] += 1

        result = []
        for i in range(n):
            # Check if current window matches
            if cnt_p == cnt_w:
                result.append(i)

            # Move the window
            cnt_w[ord(s[i]) - ord('a')] -= 1
            if i + k < n:
                cnt_w[ord(s[i + k]) - ord('a')] += 1
            else:
                break

        return result


if __name__ == "__main__":
    solution = Solution()

File: Sliding Window/leetcode_438.cpp

// Problem link ===>> https://leetcode.com/problems/find-all-anagrams-in-a-string/description/

#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define endl '\n'

class Solution
{
public:
    vector<int> findAnagrams(string s, string p)
    {
        vector<int> cnt_p(26, 0);
        vector<int> cnt_w(26, 0);

        for (char c : p)
        {
            cnt_p[c - 'a']++;
        }

        int k = p.size();
        int n = s.size();
        for (int i = 0; i < k; i++)
        {
            cnt_w[s[i] - 'a']++;
        }

        for (int i = 0; i < n; i++)
        {
            bool matched = true;
            for (int j = 0; j < 26; j++)
            {
                if (cnt_p[j] != cnt_w[j])
                {
                    matched = false;
                    break;
                }
            }

            if (matched)
            {
                cout << i << " ";
            }

            cnt_w[s[i] - 'a']--;
            if (i + k >= n)
            {
                break;
            }
            cnt_w[s[i + k] - 'a']--;
        }
        cout << endl;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    string s, p;
    cin >> s >> p;

    return 0;
}